opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java

```diff
@@ -107,61 +107,116 @@ public boolean isAggregationResponse() {
-
             ImmutableMap.Builder<String, ExprValue> builder = new ImmutableMap.Builder<>();
-            if (hit.getInnerHits() == null || hit.getInnerHits().isEmpty()) {
-              builder.putAll(docData.tupleValue());
-            } else {
-              Map<String, Object> rowSource = hit.getSourceAsMap();
-              builder.putAll(ExprValueUtils.tupleValue(rowSource).tupleValue());
-            }
-
-            metaDataFieldSet.forEach(metaDataField -> {
-              if (metaDataField.equals(METADATA_FIELD_INDEX)) {
-                builder.put(METADATA_FIELD_INDEX, new ExprStringValue(hit.getIndex()));
-              } else if (metaDataField.equals(METADATA_FIELD_ID)) {
-                builder.put(METADATA_FIELD_ID, new ExprStringValue(hit.getId()));
-              } else if (metaDataField.equals(METADATA_FIELD_SCORE)) {
-                if (!Float.isNaN(hit.getScore())) {
-                  builder.put(METADATA_FIELD_SCORE, new ExprFloatValue(hit.getScore()));
-                }
-              } else if (metaDataField.equals(METADATA_FIELD_MAXSCORE)) {
-                if (maxScore != null) {
-                  builder.put(METADATA_FIELD_MAXSCORE, maxScore);
-                }
-              } else { // if (metaDataField.equals(METADATA_FIELD_SORT)) {
-                builder.put(METADATA_FIELD_SORT, new ExprLongValue(hit.getSeqNo()));
-              }
-            });
-
-            if (!hit.getHighlightFields().isEmpty()) {
-              var hlBuilder = ImmutableMap.<String, ExprValue>builder();
-              for (var es : hit.getHighlightFields().entrySet()) {
-                hlBuilder.put(es.getKey(), ExprValueUtils.collectionValue(
-                    Arrays.stream(es.getValue().fragments()).map(
-                        t -> (t.toString())).collect(Collectors.toList())));
-              }
-              builder.put("_highlight", ExprTupleValue.fromExprValueMap(hlBuilder.build()));
-            }
+            addParsedHitsToBuilder(builder, hit);
+            addMetaDataFieldsToBuilder(builder, hit);
+            addHighlightsToBuilder(builder, hit);
             return (ExprValue) ExprTupleValue.fromExprValueMap(builder.build());
           }).iterator();
     }
   }
+
+  /**
+   * Parse response for all hits to add to builder. Inner_hits supports arrays of objects
+   * with nested type.
+   * @param builder builder to build values from response.
+   * @param hit Search hit from response.
+   */
+  private void addParsedHitsToBuilder(
+      ImmutableMap.Builder<String, ExprValue> builder,
+      SearchHit hit
+  ) {
+    if (hit.getInnerHits() == null || hit.getInnerHits().isEmpty()) {
+      builder.putAll(
+          exprValueFactory.construct(
+              hit.getSourceAsString(),
+              false
+          ).tupleValue());
+    } else {
+      ExprValue innerHits = exprValueFactory.construct(
+          new JSONObject(hit.getSourceAsMap()).toString(),
```

Comment:
this seems crazy similar to the above if condition.  Can we change this to: 
```
builder.putAll(
          exprValueFactory.construct(
              hit.getSourceAsString(),
              hit.getInnerHits() == null || hit.getInnerHits().isEmpty()
          ).tupleValue());
```

Reply:
Thanks this is definitely cleaner.